### PR TITLE
Fix floating point exception in ChainerX inferred reshape

### PR DESCRIPTION
**Pull request by hvy**
Fixes a floating point exception in reshape in ChainerX, raising an error similar to NumPy.

The changes are quite self-describing so please refer to them.

### CODE REVIEW

File: tests/chainerx_tests/unit_tests/routines_tests/test_manipulation.py [hunk 1]
```diff
@@ -209,6 +209,7 @@ def test_rollaxis_invalid(in_shape, axis, start):
     ((2, 3, 4), (3, 4, 2)),
     ((2, 3, 4), (3, -1, 2)),
     ((2, 3, 4), (3, -3, 2)),  # -3 is treated as a -1 and is valid.
+    ((2, 0, 3), (-1,)),  # Empty to inferred.
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_manipulation.py]: How about adding some more tests, like:
`(2, 0, 3) -> (-1, 4)`
`(2, 0, 3) -> (4, -1)`
`(2, 0, 3) -> (4, -1, 5)`

File: tests/chainerx_tests/unit_tests/routines_tests/test_manipulation.py [hunk 2]
```diff
@@ -351,6 +352,7 @@ def check(a_shape, b_shape):
     ((2, 3, 4), (5, -1, 3)),  # Not divisible.
     ((2, 3, 4), (-1, -1, 3)),  # More than one dimension cannot be inferred.
     ((2, 3, 4), (-2, 4, -1)),
+    ((2, 0, 4), (-1, 0)),  # Empty to high-dimensional.
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_manipulation.py]: How about adding some more tests, like:
`(2, 0, 4) -> (0, -1)`

File: tests/chainerx_tests/unit_tests/routines_tests/test_manipulation.py (hunk 1 above)
Comment (hvy) [reply tests/chainerx_tests/unit_tests/routines_tests/test_manipulation.py]: Fixed.

File: tests/chainerx_tests/unit_tests/routines_tests/test_manipulation.py (hunk 2 above)
Comment (hvy) [reply tests/chainerx_tests/unit_tests/routines_tests/test_manipulation.py]: Fixed.

**Comment by hvy** [comment]
Thanks for the suggestions, I added some more shape combinations to the tests.

**Comment by hvy** [comment]
PTAL.

**Comment by hvy** [comment]
I simplified the error message too.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by hvy** [comment]
Should we backport this?

File: chainerx_cc/chainerx/routines/manipulation.cc [hunk 3]
```diff
@@ -133,6 +133,9 @@ Shape GetInferredShape(const Shape& shape, int64_t total_size) {
         }
         int64_t rest_size = std::accumulate(inferred_shape.begin(), it, int64_t{1}, std::multiplies<>()) *
                             std::accumulate(std::next(it), inferred_shape.end(), int64_t{1}, std::multiplies<>());
+        if (rest_size == 0) {
+            throw DimensionError{"Cannot reshape array of size ", total_size, " into an ambiguous shape ", shape};
```
Comment (hvy) [inline chainerx_cc/chainerx/routines/manipulation.cc]: The shape is ambiguous if the number of dimensions is larger than 1, the total size (product) is 0, but contains a negative value to be inferred.

**Comment by niboshi** [comment]
I think we're not backporting any PRs for ChainerX, even if it's a bug fix.

**Pull request merged**